File: clojure/problem-9.clj
; A Pythagorean triplet is a set of three natural numbers, a < b < c, for which,

; a^2 + b^2 = c^2
; For example, 3^2 + 4^2 = 9 + 16 = 25 = 5^2.

; There exists exactly one Pythagorean triplet for which a + b + c = 1000.
; Find the product abc.

(def sum 1000)

(defn square [n]
  (* n n))

(defn pythagorean-triplet? [a b c]
  (true? (and (= (< a b) (< b c))
     (= (+ (square a) (square b)) (square c)))))

(first (for [a (range 1 sum) 
             b (range a (- sum a))
             :let [c (- sum a b)]
             :when (pythagorean-triplet? a b c)]
         (* a b c)))

File: go/problem9.go
// A Pythagorean triplet is a set of three natural numbers, a < b < c, for which,

// a2 + b2 = c2
// For example, 32 + 42 = 9 + 16 = 25 = 52.

// There exists exactly one Pythagorean triplet for which a + b + c = 1000.
// Find the product abc.

package main

import "fmt"

func is_pythagorean_triplet(a, b, c int) bool {
  if a < b && b < c && ((a * a) + (b * b)) == (c * c) {
    return true
  } else {
    return false
  }
}

func main() {
  limit := 1000

  for a := 1; a <= limit; a++ {
    for b := 1; b <= limit - a ; b++ {
      for c := 1; c <= limit - b; c++ {
        if is_pythagorean_triplet(a, b, c) && (a + b + c) == limit {
          fmt.Println(a * b * c)
        }
      }
    }
  }
}